code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/EdgePersonalizationResponseHandler.java

```diff
@@ -619,6 +636,82 @@ private void updateRulesEngines(
         }
     }
 
+    /**
+     * Checks to see if the user has qualified for any content cards based on provided {@link
+     * Event}.
+     *
+     * @param event may result in content card qualification.
+     */
+    void updateQualifiedContentCardsForEvent(final Event event) {
+        final Map<Surface, List<Proposition>> qualifiedContentCardsBySurface =
+                getPropositionsFromContentCardRulesEngine(event);
+        for (final Map.Entry<Surface, List<Proposition>> entry :
+                qualifiedContentCardsBySurface.entrySet()) {
+            addOrReplaceContentCards(entry.getValue(), entry.getKey());
+        }
+    }
+
+    /**
+     * Manages qualified content cards by surface. Prevents multiple entries for the same
+     * proposition in {@code contentCardsBySurface}. If an existing entry for a proposition is
+     * found, it is replaced with the value in propositions. If no prior entry exists for a
+     * proposition, a `trigger` event will be sent and written to event history).
+     *
+     * @param propositions list of qualified {@link Proposition}s for the given surface.
+     * @param surface {@link Surface} to which qualified propositions belong.
+     */
+    @SuppressWarnings("NestedIfDepth")
+    private void addOrReplaceContentCards(
+            final List<Proposition> propositions, final Surface surface) {
+        int startingCount = 0;
+        final List<Proposition> existingPropositionsArray = contentCardsBySurface.get(surface);
+        if (existingPropositionsArray != null) {
+            startingCount = existingPropositionsArray.size();
+            for (final Proposition proposition : propositions) {
+                final int existingIndex = existingPropositionsArray.indexOf(proposition);
+                if (existingIndex >= 0) {
+                    existingPropositionsArray.remove(existingIndex);
+                } else {
+                    final List<PropositionItem> propItems = proposition.getItems();
+                    final PropositionItem item = propItems.isEmpty() ? null : propItems.get(0);
+                    if (item != null) {
+                        item.track(MessagingEdgeEventType.TRIGGER);
+                    }
```

Comment:
nit: L674- L679 is repeated. Can we extract this to its own method `qualifyContentCardInProposition(final Proposition proposition)`?